File: frontend/src/types/scan-notifications.ts
/**
 * Type definitions for Producer Scan Notifications
 *
 * Real-time notification system for producers when their products are scanned.
 */

// ============================================
// Enums and Constants
// ============================================

export type NotificationChannel = 'in_app' | 'push' | 'email' | 'telegram' | 'webhook'
export type NotificationType = 'scan' | 'batch_summary' | 'first_scan' | 'suspicious' | 'milestone' | 'new_region'
export type NotificationStatus = 'pending' | 'sent' | 'delivered' | 'failed' | 'skipped'

export const CHANNEL_CONFIG: Record<NotificationChannel, { label: string; icon: string; description: string }> = {
  in_app: {
    label: 'В приложении',
    icon: 'bell',
    description: 'Уведомления в интерфейсе платформы',
  },
  push: {
    label: 'Push-уведомления',
    icon: 'smartphone',
    description: 'Мгновенные уведомления на устройство',
  },
  email: {
    label: 'Email',
    icon: 'mail',
    description: 'Уведомления на электронную почту',
  },
  telegram: {
    label: 'Telegram',
    icon: 'message-circle',
    description: 'Уведомления в Telegram бот',
  },
  webhook: {
    label: 'Webhook',
    icon: 'webhook',
    description: 'HTTP вызов внешнего сервиса',
  },
}

export const NOTIFICATION_TYPE_CONFIG: Record<NotificationType, { label: string; description: string; color: string }> = {
  scan: {
    label: 'Сканирование',
    description: 'Обычное сканирование продукта',
    color: 'bg-blue-100 text-blue-800',
  },
  batch_summary: {
    label: 'Сводка',
    description: 'Сводка сканирований за период',
    color: 'bg-gray-100 text-gray-800',
  },
  first_scan: {
    label: 'Первое сканирование',
    description: 'Первое сканирование продукта',
    color: 'bg-green-100 text-green-800',
  },
  suspicious: {
    label: 'Подозрительное',
    description: 'Подозрительная активность',
    color: 'bg-red-100 text-red-800',
  },
  milestone: {
    label: 'Достижение',
    description: 'Достигнут рубеж сканирований',
    color: 'bg-purple-100 text-purple-800',
  },
  new_region: {
    label: 'Новый регион',
    description: 'Сканирование из нового региона',
    color: 'bg-yellow-100 text-yellow-800',
  },
}

export const STATUS_CONFIG: Record<NotificationStatus, { label: string; color: string }> = {
  pending: { label: 'Ожидает', color: 'bg-yellow-500' },
  sent: { label: 'Отправлено', color: 'bg-blue-500' },
  delivered: { label: 'Доставлено', color: 'bg-green-500' },
  failed: { label: 'Ошибка', color: 'bg-red-500' },
  skipped: { label: 'Пропущено', color: 'bg-gray-500' },
}

// ============================================
// Preferences Types
// ============================================

export interface ScanNotificationPreferences {
  id: string
  organization_id: string
  enabled: boolean
  channels: NotificationChannel[]
  regions_filter: string[] | null
  notify_business_hours_only: boolean
  business_hours_start: string | null
  business_hours_end: string | null
  timezone: string
  batch_notifications: boolean
  batch_interval_minutes: number
  min_scans_for_notification: number
  notify_new_regions_only: boolean
  notify_first_scan_per_product: boolean
  notify_on_suspicious_scans: boolean
  product_ids_filter: string[] | null
  created_at: string
  updated_at: string
}

export interface ScanNotificationPreferencesUpdate {
  enabled?: boolean
  channels?: NotificationChannel[]
  regions_filter?: string[] | null
  notify_business_hours_only?: boolean
  business_hours_start?: string | null
  business_hours_end?: string | null
  timezone?: string
  batch_notifications?: boolean
  batch_interval_minutes?: number
  min_scans_for_notification?: number
  notify_new_regions_only?: boolean
  notify_first_scan_per_product?: boolean
  notify_on_suspicious_scans?: boolean
  product_ids_filter?: string[] | null
}

// ============================================
// Notification History Types
// ============================================

export interface ScanNotificationHistory {
  id: string
  organization_id: string
  scan_event_id: string | null
  channel: NotificationChannel
  notification_type: NotificationType
  status: NotificationStatus
  product_id: string | null
  batch_id: string | null
  scan_country: string | null
  scan_city: string | null
  scan_count: number
  aggregated_scan_ids: string[] | null
  error_message: string | null
  notified_at: string
  delivered_at: string | null
  metadata: Record<string, unknown> | null
}

export interface ScanNotificationHistoryListResponse {
  items: ScanNotificationHistory[]
  total: number
  page: number
  per_page: number
  total_pages: number
}

// ============================================
// Live Scan Feed Types
// ============================================

export interface LiveScanFeedItem {
  id: string
  organization_id: string
  scan_event_id: string | null
  product_id: string | null
  product_name: string | null
  product_slug: string | null
  batch_code: string | null
  country: string | null
  city: string | null
  region: string | null
  device_type: string | null
  is_first_scan: boolean
  is_suspicious: boolean
  is_new_region: boolean
  scanned_at: string
}

export interface LiveScanFeedResponse {
  items: LiveScanFeedItem[]
  total: number
  has_more: boolean
}

// ============================================
// Statistics Types
// ============================================

export interface ScanNotificationStats {
  total_notifications_sent: number
  notifications_today: number
  notifications_this_week: number
  by_channel: Record<string, number>
  by_type: Record<string, number>
  by_status: Record<string, number>
  avg_delivery_time_seconds: number | null
}

export interface LiveScanStats {
  total_scans_today: number
  total_scans_week: number
  unique_products_scanned: number
  unique_regions: number
  suspicious_scans_count: number
  first_scans_count: number
  top_countries: Array<{ country: string; count: number }>
  top_products: Array<{ product_name: string; product_id: string; count: number }>
}

// ============================================
// WebSocket Types
// ============================================

export interface WebSocketScanEvent {
  type: 'scan' | 'connected' | 'pong' | 'heartbeat' | 'error'
  data?: LiveScanFeedItem
  organization_id?: string
  message?: string
  timestamp: string
}

export type WebSocketMessageType = 'subscribe' | 'unsubscribe' | 'ping'

export interface WebSocketOutgoingMessage {
  type: WebSocketMessageType
}

// ============================================
// Form Types
// ============================================

export interface ScanNotificationSettingsFormData {
  enabled: boolean
  channels: NotificationChannel[]
  regions_filter: string[]
  notify_business_hours_only: boolean
  business_hours_start: string
  business_hours_end: string
  timezone: string
  batch_notifications: boolean
  batch_interval_minutes: number
  min_scans_for_notification: number
  notify_new_regions_only: boolean
  notify_first_scan_per_product: boolean
  notify_on_suspicious_scans: boolean
}

// ============================================
// Component Props Types
// ============================================

export interface ScanNotificationSettingsProps {
  preferences: ScanNotificationPreferences
  onSubmit: (data: ScanNotificationPreferencesUpdate) => Promise<void>
  isLoading?: boolean
}

export interface LiveScanFeedProps {
  organizationId: string
  initialItems?: LiveScanFeedItem[]
  autoRefresh?: boolean
  refreshInterval?: number
}

export interface ScanNotificationBadgeProps {
  count: number
  variant?: 'default' | 'pulse' | 'minimal'
  onClick?: () => void
}

// ============================================
// Timezone Constants
// ============================================

export const RUSSIAN_TIMEZONES = [
  { value: 'Europe/Kaliningrad', label: 'Калининград (UTC+2)' },
  { value: 'Europe/Moscow', label: 'Москва (UTC+3)' },
  { value: 'Europe/Samara', label: 'Самара (UTC+4)' },
  { value: 'Asia/Yekaterinburg', label: 'Екатеринбург (UTC+5)' },
  { value: 'Asia/Omsk', label: 'Омск (UTC+6)' },
  { value: 'Asia/Krasnoyarsk', label: 'Красноярск (UTC+7)' },
  { value: 'Asia/Irkutsk', label: 'Иркутск (UTC+8)' },
  { value: 'Asia/Yakutsk', label: 'Якутск (UTC+9)' },
  { value: 'Asia/Vladivostok', label: 'Владивосток (UTC+10)' },
  { value: 'Asia/Magadan', label: 'Магадан (UTC+11)' },
  { value: 'Asia/Kamchatka', label: 'Камчатка (UTC+12)' },
] as const

File: backend/app/schemas/scan_notifications.py
"""
Scan Notifications Schemas

Pydantic models for the real-time producer scan notifications feature.
"""

from __future__ import annotations

from datetime import datetime, time
from typing import Literal, Optional

from pydantic import BaseModel, Field


# ============================================================
# NOTIFICATION CHANNELS
# ============================================================

NotificationChannel = Literal['in_app', 'push', 'email', 'telegram', 'webhook']
NotificationType = Literal['scan', 'batch_summary', 'first_scan', 'suspicious', 'milestone', 'new_region']
NotificationStatus = Literal['pending', 'sent', 'delivered', 'failed', 'skipped']


# ============================================================
# PREFERENCES SCHEMAS
# ============================================================

class ScanNotificationPreferencesBase(BaseModel):
    """Base schema for scan notification preferences"""
    enabled: bool = True
    channels: list[NotificationChannel] = ['in_app', 'push']
    regions_filter: Optional[list[str]] = None

    # Time filters
    notify_business_hours_only: bool = False
    business_hours_start: Optional[str] = '09:00'
    business_hours_end: Optional[str] = '18:00'
    timezone: str = 'Europe/Moscow'

    # Frequency controls
    batch_notifications: bool = False
    batch_interval_minutes: int = Field(default=15, ge=1, le=60)
    min_scans_for_notification: int = Field(default=1, ge=1, le=100)

    # Filter settings
    notify_new_regions_only: bool = False
    notify_first_scan_per_product: bool = True
    notify_on_suspicious_scans: bool = True

    # Product filters
    product_ids_filter: Optional[list[str]] = None


class ScanNotificationPreferences(ScanNotificationPreferencesBase):
    """Full scan notification preferences record"""
    id: str
    organization_id: str
    created_at: datetime
    updated_at: datetime


class ScanNotificationPreferencesUpdate(BaseModel):
    """Schema for updating preferences"""
    enabled: Optional[bool] = None
    channels: Optional[list[NotificationChannel]] = None
    regions_filter: Optional[list[str]] = None
    notify_business_hours_only: Optional[bool] = None
    business_hours_start: Optional[str] = None
    business_hours_end: Optional[str] = None
    timezone: Optional[str] = None
    batch_notifications: Optional[bool] = None
    batch_interval_minutes: Optional[int] = Field(default=None, ge=1, le=60)
    min_scans_for_notification: Optional[int] = Field(default=None, ge=1, le=100)
    notify_new_regions_only: Optional[bool] = None
    notify_first_scan_per_product: Optional[bool] = None
    notify_on_suspicious_scans: Optional[bool] = None
    product_ids_filter: Optional[list[str]] = None


# ============================================================
# NOTIFICATION HISTORY SCHEMAS
# ============================================================

class ScanNotificationHistoryBase(BaseModel):
    """Base schema for notification history"""
    channel: NotificationChannel
    notification_type: NotificationType
    status: NotificationStatus
    scan_country: Optional[str] = None
    scan_city: Optional[str] = None
    scan_count: int = 1


class ScanNotificationHistory(ScanNotificationHistoryBase):
    """Full notification history record"""
    id: str
    organization_id: str
    scan_event_id: Optional[str] = None
    product_id: Optional[str] = None
    batch_id: Optional[str] = None
    error_message: Optional[str] = None
    aggregated_scan_ids: Optional[list[str]] = None
    notified_at: datetime
    delivered_at: Optional[datetime] = None
    metadata: Optional[dict] = None


class ScanNotificationHistoryListResponse(BaseModel):
    """Paginated list of notification history"""
    items: list[ScanNotificationHistory]
    total: int
    page: int
    per_page: int
    total_pages: int


# ============================================================
# LIVE SCAN FEED SCHEMAS
# ============================================================

class LiveScanFeedItem(BaseModel):
    """Single item in the live scan feed"""
    id: str
    organization_id: str
    scan_event_id: Optional[str] = None
    product_id: Optional[str] = None
    product_name: Optional[str] = None
    product_slug: Optional[str] = None
    batch_code: Optional[str] = None
    country: Optional[str] = None
    city: Optional[str] = None
    region: Optional[str] = None
    device_type: Optional[str] = None
    is_first_scan: bool = False
    is_suspicious: bool = False
    is_new_region: bool = False
    scanned_at: datetime


class LiveScanFeedResponse(BaseModel):
    """Response with live scan feed items"""
    items: list[LiveScanFeedItem]
    total: int
    has_more: bool


# ============================================================
# STATS SCHEMAS
# ============================================================

class ScanNotificationStats(BaseModel):
    """Statistics for scan notifications"""
    total_notifications_sent: int
    notifications_today: int
    notifications_this_week: int
    by_channel: dict[str, int]
    by_type: dict[str, int]
    by_status: dict[str, int]
    avg_delivery_time_seconds: Optional[float] = None


class LiveScanStats(BaseModel):
    """Statistics for live scan feed"""
    total_scans_today: int
    total_scans_week: int
    unique_products_scanned: int
    unique_regions: int
    suspicious_scans_count: int
    first_scans_count: int
    top_countries: list[dict]
    top_products: list[dict]


# ============================================================
# WEBSOCKET SCHEMAS
# ============================================================

class WebSocketScanEvent(BaseModel):
    """Real-time scan event sent via WebSocket"""
    event_type: Literal['new_scan', 'scan_update', 'notification']
    data: LiveScanFeedItem
    timestamp: datetime


class WebSocketConnectionMessage(BaseModel):
    """WebSocket connection status message"""
    type: Literal['connected', 'disconnected', 'error', 'heartbeat']
    message: Optional[str] = None
    organization_id: Optional[str] = None
    timestamp: datetime
